### ISSUE TITLE
TIME; value in HLSL shaders should match getTickCount() of Lua scripts

### ISSUE THREAD

**Issue by Einheit-101** [multitheftauto/mtasa-blue#1558]
**Description**
Currently the TIME value in HLSL shaders is a seconds counter that seems to get reset after quitting, i dont know exactly what happens.
getTickCount() is pretty reliable to compare it with the TIME value in shaders, but both values get desynced as soon as the player exits to the menu and reconnects to the server. I am pretty sure that TIME simply gets reset.

**Describe the solution you'd like**
Sync TIME with tick count *0.001 in order to have a reliable synced time between shaders and scripts. I dont think it breaks anything, most shaders use frac(time) in order to repeat certain code and the shaders that dont do this should not get any drawbacks from a slightly different TIME value. But i am no expert. 
Currently i have to loop through all shaders and manually set the current time with dxSetShaderValue.

**Comment by AlexTMjugador** [comment]
Why do you wish that the TIME value in HLSL keeps incrementing after quitting, instead of being reset? What are you doing that depends on that behavior?

**Comment by Einheit-101** [comment]
I created a shader that gets the current tick count once it gets created and then it starts doing its job depending on the time that has passed since its creation. Currently i need to loop through all shaders and pass the new current tick count instead of just using (TIME - oldTick). Maybe its a simple change to make TIME the equivalent of the tick counter. If its not, feel free to close this.

**Comment by AlexTMjugador** [comment]
For your use case, is it enough to store the first TIME value the shader sees somewhere, and then substracting the current value of TIME to calculate the elapsed time? So `Elapsed time = TIME now - first TIME`.

**Comment by Einheit-101** [comment]
But how can i define the time when the shader starts as float? I can only pass the tick count.

**Comment by qaisjp** [comment]
To summarise, are you saying that:

- `getTickCount()` returns time since user joined **first** server
- `TIME` in shader returns time since user joined server

And that you would like them to match? Seems reasonable.

**Comment by ccw808** [comment]
I haven't tested this, but:

- client `getTickCount()` calls `GetTickCount64()` in client.dll which gets loaded on connect, so `getTickCount()` should return ticks since user joined the current server
- `TIME` calls `GetTickCount64()` in core.dll, so should return seconds since MTA was started

**Comment by Einheit-101** [comment]
I guess TIME gets reset when leaving the server. Because otherwise it would be desynced when i join the first server already. I could create a simple test resource.

**Comment by Einheit-101** [comment]
> I haven't tested this, but:
> 
> * client `getTickCount()` calls `GetTickCount64()` in client.dll which gets loaded on connect, so `getTickCount()` should return ticks since user joined the current server
> * `TIME` calls `GetTickCount64()` in core.dll, so should return seconds since MTA was started

So maybe its better to call it in client.dll for shaders? If possible?

**Comment by ccw808** [comment]
> So maybe its better to call it in client.dll for shaders? If possible?

Yes. That would be the solution

**Comment by AlexTMjugador** [comment]
> But how can i define the time when the shader starts as float? I can only pass the tick count.

Good point. I don't know, and shaders have a very different execution environment compared to Lua. As of the comments above, changing the time source of TIME is seemingly easy, so that indeed sounds like the right thing to do. I can't imagine this change being used for nefarious side-channel attacks or something like that.
